common/src/main/java/com/microsoft/identity/common/adal/internal/tokensharing/TokenShareUtility.java

```diff
@@ -0,0 +1,225 @@
+                mClientId,
+                null, // wildcard (*)
+                localAccountRecord
+        );
+
+        // Inspect the result for completeness...
+        if (null == cacheRecord.getRefreshToken() || null == cacheRecord.getIdToken()) {
+            Logger.warn(
+                    TAG + methodName,
+                    "That's strange, we had an AccountRecord for OID: "
+                            + oid
+                            + " but couldn't find tokens for them."
+            );
+
+            throw new ClientException(TOKEN_CACHE_ITEM_NOT_FOUND);
+        }
+
+        final TokenCacheItem cacheItemToExport = adapt(
+                cacheRecord.getIdToken(),
+                cacheRecord.getRefreshToken()
+        );
+
+        // Ship it
+        return SSOStateSerializer.serialize(cacheItemToExport);
+    }
+
+    @Override
+    public void saveFamilyRefreshToken(@NonNull final String tokenCacheItemJson) throws Exception {
+        final Future<Pair<MicrosoftAccount, MicrosoftRefreshToken>> resultFuture =
+                sBackgroundExecutor.submit(new Callable<Pair<MicrosoftAccount, MicrosoftRefreshToken>>() {
+                    @Override
+                    public Pair<MicrosoftAccount, MicrosoftRefreshToken> call() throws ClientException {
+                        final TokenCacheItem cacheItemToRenew = SSOStateSerializer.deserialize(tokenCacheItemJson);
+
+                        // We're going to 'hijack' this token and set our own clientId for renewal
+                        // since these are FRTs, this is OK to do.
+                        cacheItemToRenew.setClientId(mClientId);
+
+                        return renewToken(mRedirectUri, cacheItemToRenew);
+                    }
+                });
+
+        final Pair<MicrosoftAccount, MicrosoftRefreshToken> resultPair = resultFuture.get();
+
+        // If an error is encountered while requesting new tokens, null is returned
+        // Check the result, before proceeding to save into the cache...
+        if (null != resultPair) {
+            mTokenCache.setSingleSignOnState(
+                    resultPair.first, // The account
+                    resultPair.second // The refresh token
+            );
+        }
+    }
+
+    @SuppressWarnings("PMD.UnusedPrivateMethod")
+    @NonNull
+    private static TokenCacheItem adapt(@NonNull final IdTokenRecord idTokenRecord,
+                                        @NonNull final RefreshTokenRecord refreshTokenRecord) throws ServiceException {
+        final TokenCacheItem tokenCacheItem = new TokenCacheItem();
+        tokenCacheItem.setClientId(refreshTokenRecord.getClientId());
+        tokenCacheItem.setAuthority(idTokenRecord.getAuthority());
+        tokenCacheItem.setRefreshToken(refreshTokenRecord.getSecret());
+        tokenCacheItem.setResource(null); // TODO Does this need to be present? You can't turn scopes into a resource AFAIK...
```

Comment:
This is my other TODO -- here I am transforming v2 credentials into a v1 equivalent.

Not clear to me what I can do about populating the resource property if everything I have in the cache is scopes based...

Reply:
I think for FOCI tokens this is null, but we can check the ADAL implementation to confirm.

Reply:
Confirming; in the examples I captured from ADAL the resource value is `null`

Reply:
Resolved in https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/529/commits/00902758e3df779e300f1a5cdbb7990ab345fa2e